### PR TITLE
fix config settings and add the rest of plot options to config

### LINKED ISSUE #302: `world_readable` config is broken

### PR DESCRIPTION
**Pull request by chriddyp**
hey @theengineear @cldougl can you take a look at this? I found the previous plot_option code super confusing to navigate, so I rewrote a bunch of it. Also, I added a ton of tests.

this fixes #302

### CODE REVIEW

**Comment by theengineear** [comment]
@chriddyp cool, got some comments for ya. I think it's good that you're looking into this refactor. It's definitely confusing and got much more so when we added some of the 'sharing' stuff in a recent pr.

Another thing that would be nice... i think we need to reorganize the python api files. here are a couple things that jump out:

* `tools.py` really shouldn't depend on graph_objs, which means we should move graph_obj-related functionality info the /graph_objs package (e.g., /graph_objs/figure_factory.py)
* getting import info for pandas and such should not require importing tools
* related to above, a circular-import-proof way to know which optional deps are imported would be great
* revisit the /plotly/plotly package in terms of keeping it restricted to network-communication logic to plot.ly

**Comment by chriddyp** [comment]
alright @theengineear - this one's for you: https://github.com/plotly/python-api/commit/4d367abdb1b749c45f9e93153908bf2864a9e02a

good to :dancer: ?

**Comment by theengineear** [comment]
Looking good! Ping me once tests pass and I'll give ya a final review?

**Comment by chriddyp** [comment]
alright @theengineear - tests are fixed.

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,12 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [Unreleased]
 
+## [1.8.5] - 2015-09-28
+- Saving "world_readable" to your config file via `plotly.tools.set_config` actually works.
+- You can also save `auto_open` and `sharing` to the config file so that you can forget these
+  keyword argument in `py.iplot` and `py.plot`.
+
+
```
Comment (theengineear) [inline CHANGELOG.md]: oops, after praising the changelog, i totally didn't add a note! gah!

Also, this should now be 1.8.6, the update hit at 1.8.5, mind giving it a quick change?

File: plotly/files.py [hunk 2]
```diff
@@ -19,7 +19,9 @@
                               'plotly_api_domain': 'https://api.plot.ly',
                               'plotly_ssl_verification': True,
                               'plotly_proxy_authorization': False,
-                              'world_readable': True}}
+                              'world_readable': True,
+                              'sharing': "public",
```
Comment (theengineear) [inline plotly/files.py]: :cow2: I think our trend is to use single quotes when possible.

File: plotly/files.py [hunk 3]
```diff
@@ -19,7 +19,9 @@
...
+                              "auto_open": True}}
```
Comment (theengineear) [inline plotly/files.py]: :cow2: see https://github.com/plotly/python-api/pull/303/files#r40791460

File: plotly/plotly/plotly.py [hunk 4]
```diff
@@ -80,68 +80,55 @@ def get_config():
...
-def get_plot_options():
+def _plot_option_logic(plot_options_from_call_signature):
     """
-    Merge default and user-defined plot options.
+    Given some plot_options as part of a plot call, decide on final options.
+    Precendence:
+        1 - Start with DEFAULT_PLOT_OPTIONS
+        2 - Update each key with ~/.plotly/.config options (tls.get_config)
+        3 - Update each key with session plot options (set by py.sign_in)
+        4 - Update each key with plot, iplot call signature options
```
Comment (theengineear) [inline plotly/plotly/plotly.py]: Great :dolls: :)

File: plotly/plotly/plotly.py [hunk 5]
```diff
@@ -80,68 +80,55 @@ def get_config():
...
+    file_options = tools.get_config_file()
+    session_options = get_session_plot_options()
+    plot_options_from_call_signature = copy.deepcopy(plot_options_from_call_signature)
+
+    # Validate options and fill in defaults w world_readable and sharing
+    for option_set in [plot_options_from_call_signature,
+                       session_options, file_options]:
+        utils.validate_world_readable_and_sharing_settings(option_set)
+        utils.set_sharing_and_world_readable(option_set)
+        # dynamic defaults
```
Comment (theengineear) [inline plotly/plotly/plotly.py]: :cow2: newline above a comment

File: plotly/plotly/plotly.py [hunk 6]
```diff
@@ -80,68 +80,55 @@ def get_config():
...
-            )
-    else:
-        current_plot_options['sharing'] = None
+    user_plot_options = default_plot_options
+    user_plot_options = update_plot_options(user_plot_options,
+                                            file_options)
+    user_plot_options = update_plot_options(user_plot_options,
+                                            session_options)
+    user_plot_options = update_plot_options(user_plot_options,
+                                            plot_options_from_call_signature)
```
Comment (theengineear) [inline plotly/plotly/plotly.py]: :cow2: Sorry, I still don't understand why we need to do the `update_plot_options` thing? It just seems like it'd be clearer to strip them out later?

```
user_plot_options = {}
user_plot_options.update(default_plot_options)
user_plot_options.update(file_options)
user_plot_options.update(session_options)
user_plot_options.update(plot_optoins_from_call_signature)
user_plot_options = {k: v for k, v in user_plot_options.items()
                     if k in default_plot_options}
```

It's basically the same number of lines and doesn't introduce anything new.

File: plotly/tests/test_core/test_plotly/test_plot.py [hunk 7]
```diff
@@ -9,70 +9,63 @@
 
 import requests
 import six
+import json
```
Comment (theengineear) [inline plotly/tests/test_core/test_plotly/test_plot.py]: :cow2: we should order the built-in `json` import above the other imports:

https://github.com/plotly/dev-docs/blob/master/reference/styleguides/python.md#imports

File: plotly/tests/test_core/test_plotly/test_plot.py [hunk 8]
```diff
@@ -9,70 +9,63 @@
...
 
 from unittest import TestCase
 from nose.plugins.attrib import attr
 from nose.tools import raises
 
+from plotly.tests.utils import PlotlyTestCase
```
Comment (theengineear) [inline plotly/tests/test_core/test_plotly/test_plot.py]: :cow2: Similar to above, we typically order `from` imports below `import` imports.

https://github.com/plotly/dev-docs/blob/master/reference/styleguides/python.md#imports

File: plotly/tests/utils.py [hunk 9]
```diff
@@ -17,6 +17,14 @@ def __init__(self, *args, **kwargs):
...
         super(PlotlyTestCase, self).__init__(*args, **kwargs)
 
+    @classmethod
+    def setUpClass(cls):
+        session._session = {
+            'credentials': {},
+            'config': {},
+            'plot_options': {}
+        }
+
```
Comment (theengineear) [inline plotly/tests/utils.py]: thx!

File: plotly/tools.py [hunk 10]
```diff
@@ -181,33 +186,45 @@ def set_config_file(plotly_domain=None,
...
     if isinstance(plotly_api_domain, six.string_types):
         settings['plotly_api_domain'] = plotly_api_domain
     elif plotly_api_domain is not None:
-        raise TypeError('Input should be a string')
+        raise TypeError('plotly_api_domain should be a string')
     if isinstance(plotly_ssl_verification, (six.string_types, bool)):
         settings['plotly_ssl_verification'] = plotly_ssl_verification
     elif plotly_ssl_verification is not None:
-        raise TypeError('Input should be a boolean')
+        raise TypeError('plotly_ssl_verification should be a boolean')
```
Comment (theengineear) [inline plotly/tools.py]: yup, better!

File: plotly/utils.py [hunk 11]
```diff
@@ -13,6 +15,9 @@
 
 import pytz
 
+
+from . exceptions import PlotlyError
```
Comment (theengineear) [inline plotly/utils.py]: :cow2: I think Python style guides still prefers absolute imports, even over explicit, relative imports.

File: plotly/utils.py [hunk 12]
```diff
@@ -399,3 +404,46 @@ def get_by_path(obj, path):
...
+            "False, sharing can only be set to 'private' or 'secret'")
+    elif ('sharing' in option_set and
+          option_set['sharing'] not in ['public', 'private', 'secret', None]):
+        raise PlotlyError(
+            "The 'sharing' argument only accepts one of the following "
+            "strings:\n'public' -- for public plots\n"
+            "'private' -- for private plots\n"
+            "'secret' -- for private plots that can be shared with a "
+            "secret url"
+        )
```
Comment (theengineear) [inline plotly/utils.py]: :cow2: This logic is hard to read though. I didn't fill out the exact messages, but would something like this work?

```
world_readable = option_set.get('world_readable')
sharing = option_set.get('sharing')

if 'world_readable' in option_set and world_readable not in [True, False]:
    raise PlotlyError('The world_readable option must be True or False.')

if 'sharing' in option_set and sharing not in ['public', 'private', 'secret']:
    raise PlotlyError("The sharing option must be 'public', 'private', or 'secret'.")

if world_readable is True and sharing != 'public':
    raise PlotlyError("If world_readable is True, sharing must be 'public'.")

if world_readable is False and sharing == 'public':
    raise PlotlyError("If world_readable is False, sharing cannot be 'public'.")
```

Just a cow tip, take it er leave it :) I didn't actually test this.

Comment (theengineear) [inline plotly/utils.py]: At minimum, can we kill the redundant check on `is None` and `is True` or `is False`. You need at most one `is` check on a value in a conditional, right?

File: plotly/version.py [hunk 13]
```diff
@@ -1 +1 @@
-__version__ = '1.8.4'
+__version__ = '1.8.5'
```
Comment (theengineear) [inline plotly/version.py]: `1.8.6` now.

File: plotly/tests/test_core/test_plotly/test_plot.py [hunk 14]
```diff
@@ -211,3 +210,156 @@ def test_private_plot_response_with_and_without_share_key(self):
 
         # The json file of the secret plot should be 200
         self.assertTrue(secret_plot_response.status_code, 200)
+
+
+class TestPlotOptionLogic(PlotlyTestCase):
+    conflicting_option_set = (
+        {'world_readable': True, 'sharing': 'secret', },
```
Comment (theengineear) [inline plotly/tests/test_core/test_plotly/test_plot.py]: pep:8ball: comma and space at the end of a dict.

**Comment by theengineear** [comment]
@chriddyp looks good!

* Please consider the :cow2:s I mentioned
* Also, just a quick update on the version number and the changelog version number

Other than that :dancer: :)

File: plotly/plotly/plotly.py (hunk 6 above)
Comment (chriddyp) [inline plotly/plotly/plotly.py]: yeah sure, either way. I'll update.

File: plotly/utils.py (hunk 11 above)
Comment (chriddyp) [inline plotly/utils.py]: this was some stuff I had to do to get around circular imports across py 27 and py 33

File: plotly/utils.py (hunk 12 above)
Comment (chriddyp) [inline plotly/utils.py]: I'd rather be explicit about the `None`. It's not obvious that these parameters would play under the rules of duck typing.

Comment (theengineear) [inline plotly/utils.py]: `option_set['world_readable'] is not None and option_set['world_readable'] is True`

How could it be `True` and `None`?

There for you can always rewrite `X is False` as `X is not None and X is False`, right?

Anyways, just a cow tip, nbd :) i could totally be off on this one too.

Comment (chriddyp) [inline plotly/utils.py]: ohh my bad, totally didn't read this all the way through, yup you're right

**Pull request merged**